src/mame/machine/bublbobl.cpp

```diff
@@ -86,41 +143,41 @@ void bublbobl_state::device_timer(emu_timer &timer, device_timer_id id, int para
 	}
 }
 
-
-WRITE8_MEMBER(bublbobl_state::bublbobl_sound_command_w)
-{
-	m_soundlatch->write(space, offset, data);
-	synchronize(TIMER_NMI, data);
-}
-
 WRITE8_MEMBER(bublbobl_state::bublbobl_sh_nmi_disable_w)
 {
-	m_sound_nmi_enable = 0;
+	m_sound_nmi_enable = false;
+	m_audiocpu->set_input_line(INPUT_LINE_NMI, CLEAR_LINE);
 }
 
 WRITE8_MEMBER(bublbobl_state::bublbobl_sh_nmi_enable_w)
 {
-	m_sound_nmi_enable = 1;
-	if (m_pending_nmi)
-	{
-		m_audiocpu->set_input_line(INPUT_LINE_NMI, PULSE_LINE);
-		m_pending_nmi = 0;
-	}
+	m_sound_nmi_enable = true;
+	if (m_main_to_sound->pending_r() && (m_sound_nmi_enable == true))
+		m_audiocpu->set_input_line(INPUT_LINE_NMI, ASSERT_LINE);
+	else
+		m_audiocpu->set_input_line(INPUT_LINE_NMI, CLEAR_LINE);
 }
 
-WRITE8_MEMBER(bublbobl_state::bublbobl_soundcpu_reset_w)
+WRITE_LINE_MEMBER(bublbobl_state::main_to_sound_cb)
 {
-	m_audiocpu->set_input_line(INPUT_LINE_RESET, data ? ASSERT_LINE : CLEAR_LINE);
+	if (state && (m_sound_nmi_enable == true))
+		m_audiocpu->set_input_line(INPUT_LINE_NMI, state);
+	else
+		m_audiocpu->set_input_line(INPUT_LINE_NMI, CLEAR_LINE);
 }
 
-READ8_MEMBER(bublbobl_state::bublbobl_sound_status_r)
+WRITE8_MEMBER(bublbobl_state::bublbobl_soundcpu_reset_w)
 {
-	return m_sound_status;
+	//logerror("soundcpu_reset_w called with data of %d\n", data);
+	common_sreset(data?ASSERT_LINE:CLEAR_LINE);
 }
 
-WRITE8_MEMBER(bublbobl_state::bublbobl_sound_status_w)
+READ8_MEMBER(bublbobl_state::common_sound_semaphores_r)
 {
-	m_sound_status = data;
+	uint8_t ret = 0xFC;
+	ret |= m_main_to_sound->pending_r()?0x2:0x0;
+	ret |= m_sound_to_main->pending_r()?0x1:0x0;
```

Comment:
I could do "if (m_main_to_sound->pending_r()) ret |= 0x2;" instead, if that's better...